### ISSUE TITLE
[REQUEST]: Track source locations in `solc`

### ISSUE THREAD

**Issue by mattgodbolt** [compiler-explorer/compiler-explorer#3520]
### Is your feature request related to a problem? Please describe

Almost every compiler emits debug info that lets us track the relationship between source and assembly. The `solc` compiler either doesn't have this or it isn't parsed correctly. e.g. https://godbolt.org/z/KTrM1nEj4 there's no coloured backgrounds showing the relationship between source and asm.

### Describe the solution you'd like

Highlighting, like other langauges.

### Describe alternatives you've considered

na

### Additional context

Not applicable

**Comment by anders-torbjornsen** [comment]
Solc can output source mappings which allow you to map asm instructions back to a code region, I've done this in the past for custom debugging/gas analysis tools and I think it should be fairly straightforward. Docs [here ](https://docs.soliditylang.org/en/latest/internals/source_mappings.html?highlight=source%20mapping), I'll try to have a go at this over the weekend.

**Comment by shekhirin** [comment]
It's not parsed correctly now. Command line options that we're using now
https://github.com/compiler-explorer/compiler-explorer/blob/2001d1b65ebef08355d52159146a9f3513b26c0f/lib/compilers/solidity.js#L45-L48
already result in outputting corresponding line numbers for assembly, e.g.:
```bash
➜  ~ echo 'pragma solidity ^0.8.9; contract Contract { function test(uint32 num) public pure returns (uint32) { return 1; } }' | solc --combined-json asm --optimize 2>/dev/null - | jq | head -n 19
{
  "contracts": {
    "<stdin>:Contract": {
      "asm": {
        ".code": [
          {
            "begin": 24,
            "end": 114,
            "name": "PUSH",
            "source": 0,
            "value": "80"
          },
          {
            "begin": 24,
            "end": 114,
            "name": "PUSH",
            "source": 0,
            "value": "40"
          },
```

---

The problem I encountered though is different: there can be multiple source code lines matching multiple assembly opcodes but if I understand correctly, it's not supported in compiler explorer now (only one source code line can be attached to one assembly opcode):
https://github.com/compiler-explorer/compiler-explorer/blob/48eee26d2c085ae548c8b084f5f418a6211bd190/lib/compilers/python.js#L51-L66

**Comment by shekhirin** [comment]
> Solc can output source mappings which allow you to map asm instructions back to a code region, I've done this in the past for custom debugging/gas analysis tools and I think it should be fairly straightforward. Docs [here ](https://docs.soliditylang.org/en/latest/internals/source_mappings.html?highlight=source%20mapping), I'll try to have a go at this over the weekend.

cool, I didn't try to use sourcemaps for that, might be useful indeed

**Comment by anders-torbjornsen** [comment]
Yeah indeed there are parts of asm which can get attributed to large chunks of source (the whole contract sometimes), iirc this is things like the generated code for routing selector codes to the right function. I'm hoping we can just ignore those

**Comment by mattgodbolt** [comment]
> only one source code line can be attached to one assembly opcode

That seems to be the case. We support a source line corresponding to multiple (disjoint) assembly instructions (like in https://godbolt.org/z/saqnvjdcn where the two `imul` instructions map back to line 3 of the source), but not the other way around.  Support could be added though.

**Comment by anders-torbjornsen** [comment]
Making some progress
![FQt7mj1X0AcBEXH](https://user-images.githubusercontent.com/95024805/164042599-39030ad0-c7f5-4b91-a493-8cd8c025aa7a.jpg)

**Comment by RubenRBS** [comment]
Nice! Let us know if you have any questions regarding the codebase

**Comment by anders-torbjornsen** [comment]
Some more progress, just need to get the tag locations some more useful names and then I can start pulling it all together
![image](https://user-images.githubusercontent.com/95024805/164790706-37ba9f67-fe13-43f2-b9ae-bfd0360653dd.png)

**Comment by anders-torbjornsen** [comment]
PR created, let me know your thoughts
